lib/Sema/CSGen.cpp

```diff
@@ -2262,47 +2305,93 @@ namespace {
         // ascribed type.
         Type subPatternType =
             getTypeForPattern(
-               cast<TypedPattern>(pattern)->getSubPattern(), locator);
+               cast<TypedPattern>(pattern)->getSubPattern(), locator,
+               Type(), bindPatternVarsOneWay);
         CS.addConstraint(
             ConstraintKind::Conversion, subPatternType, openedType, locator);
         return setType(openedType);
       }
 
       case PatternKind::Tuple: {
         auto tuplePat = cast<TuplePattern>(pattern);
+
+        // If there's an externally-imposed type, decompose it into element
+        // types so long as we have the right number of such types.
+        SmallVector<AnyFunctionType::Param, 4> externalEltTypes;
```

Comment:
I see that `externalEltTypes` are only used to retrieve types, maybe it would be easier to just store types directly instead of `AnyFunctionType::Param`? Also it seems weird to use `decomposeInput` in this context since it has nothing to do with function types besides `externalPatternType` possibly being a `TupleType`...

Reply:
I'm using `decomposeInput` because patterns are handled like calls, with "paren" and "tuple" variants. We should change that, but for now `decomposeInput` is the most direct way to deal with splitting apart the paren-or-tuple types we get for the arguments.

Reply:
Ok, that makes sense, I just thought that it's always going to be a `TupleType`.